guides/source/action_controller_advanced_topics.md

```diff
@@ -0,0 +1,500 @@
+  <%= form.text_field :username %>
+  <%= form.text_field :password %>
+<% end %>
+```
+
+A CSRF token named `authenticity_token` is added as a hidden field in the generated HTML:
+
+```html
+<form accept-charset="UTF-8" action="/users/1" method="post">
+<input type="hidden"
+       value="67250ab105eb5ad10851c00a5621854a23af5489"
+       name="authenticity_token"/>
+<!-- fields -->
+</form>
+```
+
+Rails adds this token to every `form` that's generated using the [form helpers](form_helpers.html), so most of the time you don't need to do anything. If you're writing a form manually or need to add the token for another reason, it's available through the [`form_authenticity_token`](https://api.rubyonrails.org/classes/ActionController/RequestForgeryProtection.html#method-i-form_authenticity_token) method.
+
+```html
+<!-- app/views/layouts/application.html.erb -->
+<head>
+  <meta name="csrf-token" content="<%= form_authenticity_token %>">
+</head>
+```
+
+The `form_authenticity_token` method generates a valid authentication token. That can be useful in places where Rails does not add it automatically, like in custom Ajax calls.
+
+NOTE: All subclasses of `ActionController::Base` are protected by default and will raise an `ActionController::InvalidAuthenticityToken` error on unverified requests.
+
+You can learn more details about the CSRF attack as well as CSRF countermeasures in the [Security Guide](security.html#cross-site-request-forgery-csrf).
+
+HTTP Authentication
+-------------------
+
+Rails comes with three built-in HTTP authentication mechanisms:
+
+* Basic Authentication
+* Digest Authentication
+* Token Authentication
+
+### HTTP Basic Authentication
+
+HTTP Basic Authentication is a simple authentication method where a user is required to enter a username and password to access a website or a particular section of a website (e.g. admin section). These credentials are entered into a browser's HTTP basic dialog window. The user’s credentials are then encoded and sent in the HTTP header with each request.
+
+HTTP basic authentication is an authentication scheme that is supported by most browsers. Using HTTP Basic authentication in a Rails controller can be done by using the [`http_basic_authenticate_with`][] method:
+
+```ruby
+class AdminsController < ApplicationController
+  http_basic_authenticate_with name: "Arthur", password: "42424242"
+end
+```
+
+With the above in place, you can create controllers that inherit from
+`AdminsController`. All actions in those controllers will use HTTP basic
+authentication and require user credentials.
+
+WARNING: HTTP Basic Authentication is easy to implement but less secure on its
+own, as it will send unencrypted credentials over the network. Make sure to use
+HTTPS when using Basic Authentication. [`http_basic_authenticate_with`]:
+https://api.rubyonrails.org/classes/ActionController/HttpAuthentication/Basic/ControllerMethods/ClassMethods.html#method-i-http_basic_authenticate_with
```

Comment:
As @havenwood noticed, I think this should be:
```suggestion
HTTPS when using Basic Authentication. [`http_basic_authenticate_with`](
https://api.rubyonrails.org/classes/ActionController/HttpAuthentication/Basic/ControllerMethods/ClassMethods.html#method-i-http_basic_authenticate_with)
```

Reply:
Fixed! (A newline was missing. This one is the alternate link syntax)